Sound generation

aga11313 opened the pull request:
Commits with restrucurised sound generation parts of the application.
Introduces new classes to impove the structure.
Changes the hierarchy of the exisiting classes.
Improves the music generation algorithms.

aga11313 committed:
Restructure the Sound Generation class. Change class hierarchy.

Sound generation class was restructured and divided into a seperate class
for Channel sound generation. Classes such as instrument and channel
 were added to imporve the structure and readability of the code.

aga11313 committed:
Restructure scale generation methods.

Call scale generation methods from a Channel constructor.
Store the values as variebles of Channel.

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -0,0 +1,177 @@
+     * @return - return an Instrument object from instrumentList
+     */
+    public static Instrument generateRandomInstrument(Random random){
+        int instrumentNumber = random.nextInt((6 - 0) + 1) + 0;
```
Getting a random element from an array should use the array's length as opposed to hardcoded numbers

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Channel.java:
```diff
@@ -8,39 +8,31 @@
-    int range_max;
+    Instrument instrument;
+    Random randomGenerator;
+    //String scaleName; //major or minor
```
Commented out code should not be committed

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -0,0 +1,177 @@
+
+        int timestampCounter = 0;
+
+        StartingSequence instrument_change = new StartingSequence(192 + c.number, c.instrument, timestampCounter);
```
What is the significance of 192?

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -0,0 +1,177 @@
+         * TODO: implement the way of setting the length of the track differently (ie. a number of
+         * bars so that tunes are synchronised)
+         */
+        for (int i = 0; i < 16; i = i+2) {
```
Why i+2 when i isn't being used in the loop, could this be simplified? Also, ideally i should be named more descriptively.

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -0,0 +1,177 @@
+     * @return - a total starting code value
+     */
+    public static int generateStartingCode(int channelNumber){
+        return 144 + channelNumber;
```
What exactly is 144? It's good that this is in its own documented function though.

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -0,0 +1,177 @@
+     * @return - a value between 0 and 127
+     */
+    public static int generatePitch(Channel c){
+        int randomGeneratedPitch = c.scaleValues[c.randomGenerator.nextInt((7 - 0) + 1)];
```
I think using c.scaleValues.length would be clearer than (7-0)+1.

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -0,0 +1,177 @@
+
+public class ChannelSoundGeneration {
+
+    int scale;
```
Every function in this class is static, are these variables actually used anywhere? If so, consider just using them in the function that uses them instead of declaring them non-static out here.

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -0,0 +1,177 @@
+     * @param emotion - an emotion name
+     * @return - an array of steps for an appropriate scale
+     */
+    public static int[] chooseScaleStep(String emotion) {
```
Maybe restructure this to be something more like

```
switch (emotion) {
    case "happy":
        return new int[] {steps...};
    case "sad":
        return new int[] {steps...};
etc
}
```
to avoid instantiating unused arrays every time

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Instrument.java:
```diff
@@ -0,0 +1,28 @@
+    int rangeMin;
+    int rangeMax;
+
+    public static String[] instrument_list = new String[] {"piano", "guitar", "trombone", "trumpet", "violin", "saxophone", "flute"};
```
These are now in ChannelSoundGeneration, they shouldn't need to be duplicated

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/MidiSequence.java:
```diff
@@ -0,0 +1,16 @@
+ * Created by aga on 02/02/17.
+ */
+
+public class MidiSequence {
```
Some comment saying this is the superclass for Note and StartingSequence

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Sound.java:
```diff
@@ -18,30 +19,26 @@
 
 public class Sound extends AppCompatActivity implements MidiDriver.OnMidiStartListener {
```
More class Javadoc pls, also why does it implement Midi.OnMidiStartListener

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Sound.java:
```diff
@@ -18,30 +19,26 @@
 
     protected long userId = MainActivity.uuid_long;
-    protected long randomSeed = userId;
+    //protected long randomSeed = userId;
```
Again, commented out code. This code exists in the previous commit, so if you really needed it again, you could pull it from there.

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Sound.java:
```diff
@@ -18,30 +19,26 @@
     protected MidiDriver midi;
-    protected MediaPlayer player;
 
     protected long userId = MainActivity.uuid_long;
```
This is not the way to go about it. Use BluetoothHelper.getDeviceUUID

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Sound.java:
```diff
@@ -123,11 +141,11 @@ public void onMidiStart() {
+    /**
+     * A safe sleep function including a try catch statement to capture exceptions.
+     * Created to avoid using try catch every time a Thread sleeps.
+     * @param time - a sleep time
```
A sleep time in what units? Also, what exceptions does it capture and why?

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Sound.java:
```diff
@@ -136,114 +154,34 @@ public void safeSleep(int time) {
-    }
+        public void run() {
+            int currentTimestamp = 0;
 
```
Your spacing is inconsistent (number of newlines, number of spaces between things etc)

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/SoundGeneration.java:
```diff
@@ -10,135 +18,88 @@
 
 public class SoundGeneration {
```
Why is this class still a thing? From what I can see, all methods are static. This should be combined with ChannelSoundGeneration I think

aga11313 replied:
I think that the distinction between SoundGeneration and ChannelSoundGeneration is important as the first one takes care of everything to do with mergin and sorting MIDI sequence files and the sedond is responsible for generation of MidiSequences for one Channel Specifically.

notexactlyawe reviewed (changes requested):
All in all, this is a lot cleaner and I like the direction you have taken the code. There are however a few comments that I would like to either be defended or addressed.

aga11313 committed:
Cleanup code after restructure.

Adress the review submitted after a pull request on recent
restructure changes was requested.
Minor changes to code (not in structure) attempting to make
it clearer, cleaner and more consistent.

aga11313 commented:
There has been another commit that includes changes adressing your comments.

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -26,13 +26,22 @@
      */
     public static Instrument generateRandomInstrument(Random random){
-        int instrumentNumber = random.nextInt((6 - 0) + 1) + 0;
+        int instrumentNumber = random.nextInt((instrumentList.length - 0) + 1) + 0;
```
You don't need length - 0 +1
nextInt will generate an int in range (0, length], meaning you will get a valid index with just that. I think this will cause an occasional OutOfBoundsException
`int instrumentNumber = random.nextInt(instrumentList.length);`

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -46,26 +55,28 @@ public static Instrument generateRandomInstrument(Random random){
          */
-        for (int i = 0; i < 16; i = i+2) {
-            int startingCode = generateStartingCode(c.number);
+        for (int noteCounter = 0; noteCounter < CHANNEL_NOTE_AMOUNT; noteCounter += 2) {
```
Still using += 2

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -89,7 +100,8 @@ public static int generateStartingCode(int channelNumber){
      */
     public static int generatePitch(Channel c){
-        int randomGeneratedPitch = c.scaleValues[c.randomGenerator.nextInt((7 - 0) + 1)];
+        int arrayIndex = c.randomGenerator.nextInt((c.scaleValues.length) + 1);
```
No more +1, out of bounds

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/ChannelSoundGeneration.java:
```diff
@@ -143,17 +155,19 @@ public static  int chooseStartingNote (Instrument instrument, Random randomGener
-        int[] step_major = new int[] {2,2,1,2,2,2,1};
-        int[] step_minor = new int[] {2,1,2,2,1,2,2};
 
         int[] step_array;
```
No longer necessary I don't think

notexactlyawe commented on app/src/main/java/com/hillnerds/soundsparrow/Sound.java:
```diff
@@ -16,14 +16,13 @@
+ * A class responsible for synthesizing the MIDI Sequence files produced by SoundGeneration and
+ * ChannelSoundGeneration
+ */
 public class Sound extends AppCompatActivity implements MidiDriver.OnMidiStartListener {
```
Still don't understand the implements

notexactlyawe reviewed (commented):
Much cleaner, once you have made sure the code runs I would be happy for this to merge

aga11313 committed:
Address changes requested in PR #10 + others

Change flute definition
Make MIDI synthesise forever
